man/wasm2wat.1
.Dd $Mdocdate$
.Dt WABT 1
.Os
.Sh NAME
.Nm wasm2wat
.Nd translate from the binary format to the text format
.Sh SYNOPSIS
.Nm wasm2wat
.Op options
.Ar filename
.Sh DESCRIPTION
.Nm
Read a file in the WebAssembly binary format, and convert it to  the WebAssembly text format.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl Fl help
Print this help message
.It Fl Fl version
Print version information
.It Fl v , Fl Fl verbose
Use multiple times for more info
.It Fl o , Fl Fl output=FILENAME
Output file for the generated wast file, by default use stdout
.It Fl f , Fl Fl fold-exprs
Write folded expressions where possible
.It Fl Fl disable-exceptions
Disable Experimental exception handling
.It Fl Fl disable-mutable-globals
Disable Import/export mutable globals
.It Fl Fl disable-saturating-float-to-int
Disable Saturating float-to-int operators
.It Fl Fl disable-sign-extension
Disable Sign-extension operators
.It Fl Fl disable-simd
Disable SIMD support
.It Fl Fl enable-threads
Enable Threading support
.It Fl Fl enable-function-references
Enable Typed function references
.It Fl Fl disable-multi-value
Disable Multi-value
.It Fl Fl disable-tail-call
Disable Tail-call support
.It Fl Fl disable-bulk-memory
Disable Bulk-memory operations
.It Fl Fl disable-reference-types
Disable Reference types (externref)
.It Fl Fl disable-annotations
Disable Custom annotation syntax
.It Fl Fl enable-code-metadata
Enable Code metadata
.It Fl Fl enable-gc
Enable Garbage collection
.It Fl Fl disable-memory64
Disable 64-bit memory
.It Fl Fl disable-multi-memory
Disable Multi-memory
.It Fl Fl disable-extended-const
Disable Extended constant expressions
.It Fl Fl disable-relaxed-simd
Disable Relaxed SIMD
.It Fl Fl enable-custom-page-sizes
Enable Custom page sizes
.It Fl Fl enable-compact-imports
Enable Compact import section
.It Fl Fl enable-wide-arithmetic
Enable Wide arithmetic
.It Fl Fl enable-all
Enable all features
.It Fl Fl inline-exports
Write all exports inline
.It Fl Fl inline-imports
Write all imports inline
.It Fl Fl no-debug-names
Ignore debug names in the binary file
.It Fl Fl ignore-custom-section-errors
Ignore errors in custom sections
.It Fl Fl generate-names
Give auto-generated names to non-named functions, types, etc.
.El
.Sh EXAMPLES
parse binary file test.wasm and write text file test.wast
.Pp
.Dl $ wasm2wat test.wasm -o test.wat
parse test.wasm, write test.wat, but ignore the debug names, if any
.Pp
.Dl $ wasm2wat test.wasm --no-debug-names -o test.wat
.Sh SEE ALSO
.Xr wasm-interp 1 ,
.Xr wasm-objdump 1 ,
.Xr wasm-stats 1 ,
.Xr wasm-strip 1 ,
.Xr wasm-validate 1 ,
.Xr wasm2c 1 ,
.Xr wast2json 1 ,
.Xr wat-desugar 1 ,
.Xr wat2wasm 1 ,
.Xr spectest-interp 1 
.Sh BUGS
If you find a bug, please report it at
.br
.Lk https://github.com/WebAssembly/wabt/issues .


src/tools/wasm2wat.cc
/*
 * Copyright 2016 WebAssembly Community Group participants
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <cassert>
#include <cinttypes>
#include <cstdio>
#include <cstdlib>

#include "wabt/apply-names.h"
#include "wabt/binary-reader-ir.h"
#include "wabt/binary-reader.h"
#include "wabt/error-formatter.h"
#include "wabt/feature.h"
#include "wabt/generate-names.h"
#include "wabt/ir.h"
#include "wabt/option-parser.h"
#include "wabt/stream.h"
#include "wabt/validator.h"
#include "wabt/wast-lexer.h"
#include "wabt/wat-writer.h"

using namespace wabt;

static std::string s_infile;
static std::string s_outfile;
static Features s_features;
static bool s_generate_names;
static bool s_fold_exprs;
static bool s_inline_import;
static bool s_inline_export;
static bool s_read_debug_names = true;
static bool s_fail_on_custom_section_error = true;
static std::unique_ptr<FileStream> s_log_stream;
static bool s_validate = true;

static const char s_description[] =
    R"(  Read a file in the WebAssembly binary format, and convert it to
  the WebAssembly text format.

examples:
  # parse binary file test.wasm and write text file test.wast
  $ wasm2wat test.wasm -o test.wat

  # parse test.wasm, write test.wat, but ignore the debug names, if any
  $ wasm2wat test.wasm --no-debug-names -o test.wat
)";

static void ParseOptions(int argc, char** argv) {
  OptionParser parser("wasm2wat", s_description);

  parser.AddOption('v', "verbose", "Use multiple times for more info",
                   []() { s_log_stream = FileStream::CreateStderr(); });
  parser.AddOption(
      'o', "output", "FILENAME",
      "Output file for the generated wast file, by default use stdout",
      [](const char* argument) {
        s_outfile = argument;
        ConvertBackslashToSlash(&s_outfile);
      });
  parser.AddOption('f', "fold-exprs", "Write folded expressions where possible",
                   []() { s_fold_exprs = true; });
  s_features.AddOptions(&parser);
  parser.AddOption("inline-exports", "Write all exports inline",
                   []() { s_inline_export = true; });
  parser.AddOption("inline-imports", "Write all imports inline",
                   []() { s_inline_import = true; });
  parser.AddOption("no-debug-names", "Ignore debug names in the binary file",
                   []() { s_read_debug_names = false; });
  parser.AddOption("ignore-custom-section-errors",
                   "Ignore errors in custom sections",
                   []() { s_fail_on_custom_section_error = false; });
  parser.AddOption(
      "generate-names",
      "Give auto-generated names to non-named functions, types, etc.",
      []() { s_generate_names = true; });
  parser.AddOption("no-check", "Don't check for invalid modules",
                   []() { s_validate = false; });
  parser.AddArgument("filename", OptionParser::ArgumentCount::One,
                     [](const char* argument) {
                       s_infile = argument;
                       ConvertBackslashToSlash(&s_infile);
                     });
  parser.Parse(argc, argv);
}

int ProgramMain(int argc, char** argv) {
  Result result;

  InitStdio();
  ParseOptions(argc, argv);

  std::vector<uint8_t> file_data;
  result = ReadFile(s_infile.c_str(), &file_data);
  if (Succeeded(result)) {
    Errors errors;
    Module module;
    const bool kStopOnFirstError = true;
    ReadBinaryOptions options(s_features, s_log_stream.get(),
                              s_read_debug_names, kStopOnFirstError,
                              s_fail_on_custom_section_error);
    result =
        ReadBinaryIr(s_infile.c_str(), file_data, options, &errors, &module);
    if (Succeeded(result)) {
      if (Succeeded(result) && s_validate) {
        ValidateOptions options(s_features);
        result = ValidateModule(&module, &errors, options);
      }

      if (Succeeded(result) && s_generate_names) {
        result = GenerateNames(&module);
      }

      if (Succeeded(result)) {
        /* TODO(binji): This shouldn't fail; if a name can't be applied
         * (because the index is invalid, say) it should just be skipped. */
        Result dummy_result = ApplyNames(&module);
        WABT_USE(dummy_result);
      }

      if (Succeeded(result)) {
        WriteWatOptions wat_options(s_features);
        wat_options.fold_exprs = s_fold_exprs;
        wat_options.inline_import = s_inline_import;
        wat_options.inline_export = s_inline_export;
        FileStream stream(!s_outfile.empty() ? FileStream(s_outfile)
                                             : FileStream(stdout));
        result = WriteWat(&stream, &module, wat_options);
      }
    }
    FormatErrorsToFile(errors, Location::Type::Binary);
  }
  return result != Result::Ok;
}

int main(int argc, char** argv) {
  WABT_TRY
  return ProgramMain(argc, argv);
  WABT_CATCH_BAD_ALLOC_AND_EXIT
}
